### PR TITLE
[memory] Statically Dispatched Buffer

### PR DESCRIPTION
**Pull request by ppenna**
Description
=======

**This is a draft PR**.

I have opened it to ease discussing on the _initial implementation_ of a statically dispatched buffer abstraction.

Note that:

- I am ok on changing the name of whatever function / field. I currently just rearranged existing structures ([DataBuffer ](https://github.com/demikernel/runtime/blob/dev/src/memory/databuffer.rs) and [DPDKBuffer](https://github.com/demikernel/runtime/blob/dev/src/memory/databuffer.rs) (renamed)).
- There is a parent enumeration called `Buffer` that is used to dispatch calls to the child structures (ie. DataBuffer and DPDKBuffer)
- Implementations of constructor-like functions (ie: `new()`, `from_raw()`, `from_slice()` …) and destructors (ie `drop()`) are present only on the child structures. The rationale for this is because even with static dispatch, the responsibility for creating the underlying buffer should be from whoever manipulates Buffer structures.
- We can add an `EmptyBuffer` representation and provide a `empty()` operation in `Buffer`.

### CODE REVIEW

**Comment by BrianZill** [comment]
I had envisioned this as "everything is a DataBuffer", with no parent "Buffer" type.  Everything supporting different allocators would then be encapsulated in the DataBuffer type.  The DataBuffer type would have an internal enum field representing which allocator is providing the backing store (by "representing", I mean that the enum would actually be a number (some small integer) in some number space we define like "0 = Rust default", "1 = DPDK", "2 = Whatever").

There are trade-offs here, of course.  The positive would be that we could share parts or the entirety of some member functions between the DataBuffers backed by different allocators.  It would discourage the creation of allocator-specific member functions.  And we would avoid the extra match check on every member function call.  The negative is that we have to have explicit allocator checks inside the constructor/destructor member functions, but those should be less than for every function.

Note that I'm thinking of the difference between buffer "types" to be purely where the memory was allocated from.  I would expect, for example, that we could use the same adjust() and trim() implementations for both the Rust-default allocator and the DPDK allocator, rather than calling rte_pktmbuf_adj() and rte_pktmbuf_trim().

Eventually I want to add code to provide link-list functionality, and that should be shareable across all allocators.  So I'd want to just have one implementation of that on the one-and-only DataBuffer type.

**Comment by ppenna** [comment]
@BrianZill, unfortunately, DPDK imposes some additional requirements on how we can manipulate `mbufs`.

For instance, for adjusting and trimming, we should rely on the API that it is exposed.

Concerning the way that you thought about having a single `Buffer` structure and simply relying on a different backend implementation for the memory allocator, could you express in a code snippet what do you aim to achieve? For instance, in the current API, this is how it looks like:

```
let bufA: Buffer = Buffer::Heap(DataBuffer::new(32));   // Instantiates a 32-byte heap-managed buffer.
let bufB: Buffer = Buffer::DPDK(DPDKBuffer::new(32));   // Instantiates a 32-byte heap-managed buffer.

// Manipulate buffers.
bufA.trim(4);
bufB.adjust(8);
```

Also, could you present a draft on this `Buffer` structure that you envision?

**Comment by BrianZill** [comment]
@ppenna, we only need to use the DPDK routines for adjusting/trimming if we want to hand the buffer to something else that expects DPDK buffers.  And then we can adjust/trim as needed as part of the conversion process.  So only the send side would need to convert back.

In your code snippet, did you mean the second "Instantiates a 32-byte heap-managed buffer" to be "Instantiates a 32-byte DPDK-managed buffer"?

Basically, I just want to be able to call all the fast-path (commonly used) methods without having to execute any more instructions than absolutely necessary.  So I'd like these common methods to have the same implementation regardless of where the underlying memory came from.  Really only alloc/~~free~~drop and the to/from conversion functions should care what the allocator is.

**Comment by ppenna** [comment]
@BrianZill 

> In your code snippet, did you mean the second "Instantiates a 32-byte heap-managed buffer" to be "Instantiates a 32-byte DPDK-managed buffer"?

Yes, that is what I mean. Sorry fot the typo, I've fixed it.

> Basically, I just want to be able to call all the fast-path (commonly used) methods without having to execute any more instructions than absolutely necessary. So I'd like these common methods to have the same implementation regardless of where the underlying memory came from. Really only alloc/freedrop and the to/from conversion functions should care what the allocator is.

I understand and share the same idea as you, but I have figured out that what you are trying to achieve here in a Rust way, I couldn't figure out how. Please, provide a Rust snippet that illustrates your thoughts on how to do this.

**Comment by BrianZill** [comment]
As we discussed earlier, it seems best to proceed with this as is, which is to say, achieve polymorphism via Rust enum, rather than shoehorn everything into a single data type.  This is more flexible in that we can try out different buffer types more easily.  We could also convert into a single type if we end up deciding that is preferable.

**Comment by ppenna** [comment]
> As we discussed earlier, it seems best to proceed with this as is, which is to say, achieve polymorphism via Rust enum, rather than shoehorn everything into a single data type. This is more flexible in that we can try out different buffer types more easily. We could also convert into a single type if we end up deciding that is preferable.

Thanks @BrianZill .

I've enabled review on this PR.

**Review by BrianZill — APPROVED** [review]
Looks good, let's run with this and see how it works in practice.
